src/code/code_800EC960.c

```diff
@@ -1203,916 +1233,1078 @@ u8 D_8016B9F4;
+                                   &D_801333E8);
+        } else if ((sPrevOcarinaPitch != OCARINA_PITCH_NONE) && (sCurOcarinaPitch == OCARINA_PITCH_NONE)) {
+            // cancels ocarina for a non-valid input
             Audio_StopSfxById(NA_SE_OC_OCARINA);
         }
     }
 }
 
-void func_800ED848(u8 inputEnabled) {
-    sOcarinaInpEnabled = inputEnabled;
+/**
+ * Directly enable the ocarina to receive input without
+ * properly resetting it based on an ocarina instrument id
+ * Unused.
+ */
+void AudioOcarina_EnableInput(u8 inputEnabled) {
+    sIsOcarinaInputEnabled = inputEnabled;
 }
 
-void Audio_OcaSetInstrument(u8 arg0) {
-    if (D_80130F10 == arg0) {
+/**
+ * Resets ocarina properties based on the ocarina instrument id
+ * If ocarina instrument id is "OCARINA_INSTRUMENT_OFF", turn off the ocarina
+ * For all ocarina instrument ids, turn the ocarina on with the instrument id
+ */
+void AudioOcarina_SetInstrument(u8 ocarinaInstrumentId) {
+    if (sOcarinaInstrumentId == ocarinaInstrumentId) {
         return;
     }
 
-    Audio_SeqCmd8(SEQ_PLAYER_SFX, 1, SFX_PLAYER_CHANNEL_OCARINA, arg0);
-    D_80130F10 = arg0;
-    if (arg0 == 0) {
-        sCurOcarinaBtnPress = 0;
-        sPrevOcarinaBtnPress = 0;
-        D_8016BA18 = 0;
-        D_8016BA10 = 0xFFFF;
-        func_800ED458(0);
+    Audio_SeqCmd8(SEQ_PLAYER_SFX, 1, SFX_CHANNEL_OCARINA, ocarinaInstrumentId);
+    sOcarinaInstrumentId = ocarinaInstrumentId;
+    if (ocarinaInstrumentId == OCARINA_INSTRUMENT_OFF) {
+        sOcarinaInputButtonCur = 0;
+        sOcarinaInputButtonPrev = 0;
+        sOcarinaInputButtonPress = 0;
+
+        sOcarinaInputButtonStart = 0xFFFF;
+
+        AudioOcarina_PlayControllerInput(false);
         Audio_StopSfxById(NA_SE_OC_OCARINA);
         Audio_SetSoundBanksMute(0);
         sPlaybackState = 0;
-        sStaffPlaybackPos = 0;
-        sOcarinaInpEnabled = 0;
-        D_80130F3C = 0;
-        Audio_ClearBGMMute(SFX_PLAYER_CHANNEL_OCARINA);
+        sPlaybackStaffPos = 0;
+        sIsOcarinaInputEnabled = false;
+        sOcarinaFlags = 0;
+        // return to full volume for players 0 and 3 (background bgm) after ocarina is finished
+        Audio_ClearBGMMute(SFX_CHANNEL_OCARINA);
     } else {
-        sCurOcarinaBtnPress = 0;
-        Audio_GetOcaInput();
-        D_8016BA10 = sCurOcarinaBtnPress;
-        Audio_QueueSeqCmdMute(SFX_PLAYER_CHANNEL_OCARINA);
+        sOcarinaInputButtonCur = 0;
+        AudioOcarina_ReadControllerInput();
+        // Store button used to turn on ocarina
+        sOcarinaInputButtonStart = sOcarinaInputButtonCur;
+        // lowers volumes of players 0 and 3 (background bgm) while playing ocarina
+        Audio_QueueSeqCmdMute(SFX_CHANNEL_OCARINA);
     }
 }
 
-void Audio_OcaSetSongPlayback(s8 songIdxPlusOne, s8 playbackState) {
-    if (songIdxPlusOne == 0) {
+void AudioOcarina_SetPlaybackSong(s8 songIndexPlusOne, s8 playbackState) {
```

Comment:
I think `playbackState` is the "repeat count" of the song, based on the code neighbouring `sPlaybackState--;` in `AudioOcarina_PlaybackSong`

Reply:
Yeah, each of these different `OcarinaStaff` variables use the "state" part in completely different ways. `playbackState` seems to be always set to 1, except for teaching songs, in which it's set to 2. Interesting

Reply:
It's possible, the main reason is that there's 3 ocarinaStaff's:
`sPlayingStaff`
`sPlaybackStaff`
`sRecordingStaff`
```
typedef struct {
    /* 0x0 */ u8 buttonIndex;
    /* 0x1 */ u8 state;
    /* 0x2 */ u8 pos;
} OcarinaStaff; // size = 0x3
```
And `sPlaybackStaff.state = sPlaybackState`, it's being saved to the state of the Playback Staff. So using state is consistent with `sRecordingState` and the function `AudioOcarina_GetPlayingState`. It's always getting the state of the OcarinaStaff.

The annoying part is that even though it's the same struct for all 3 staffs, the states represents something different for each of the 3 staffs.

Reply:
Why are the three the same struct? because z_message stores all of them to the same place? Is there any chance z_message could be using a union maybe?

Reply:
Yeah:
```
msgCtx->ocarinaStaff = AudioOcarina_GetPlayingStaff()
msgCtx->ocarinaStaff = AudioOcarina_GetPlaybackStaff()
msgCtx->ocarinaStaff = AudioOcarina_GetRecordingStaff()
```
Search:
`msgCtx->ocarinaStaff->state` in the repo to see how it's all over the place. Maybe a union could make sense here? This variable seems to be kinda a mess by the look of things

Reply:
`state` seems good enough, just a bit sad it can't be more specific. Even for a count it's decent enough, I'm okay with leaving it be for now

Reply:
I'll add a comment where ocarinaStaff is defined expanding on `state` and its different meanings